azext_iot/iothub/providers/state.py

```diff
@@ -0,0 +1,692 @@
+# coding=utf-8
+# --------------------------------------------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License. See License.txt in the project root for license information.
+# --------------------------------------------------------------------------------------------
+
+from knack.log import get_logger
+
+from azext_iot.common.shared import DeviceAuthApiType, DeviceAuthType, ConfigType
+from azext_iot.iothub.common import (
+    IMMUTABLE_AND_DUPLICATE_MODULE_TWIN_FIELDS,
+    IMMUTABLE_DEVICE_IDENTITY_FIELDS,
+    IMMUTABLE_MODULE_IDENTITY_FIELDS,
+    HubAspects
+)
+from azext_iot.iothub.providers.base import IoTHubProvider
+from azext_iot.common._azure import (
+    parse_iot_hub_message_endpoint_connection_string,
+    parse_storage_container_connection_string,
+    # parse_cosmos_db_connection_string
+)
+from azure.cli.core.azclierror import (
+    FileOperationError, ResourceNotFoundError, BadRequestError, AzCLIError, RequiredArgumentMissingError
+)
+from azext_iot.common.embedded_cli import EmbeddedCLI
+from azext_iot.operations.hub import (
+    _iot_device_set_parent,
+    _iot_device_show,
+    _iot_device_module_twin_show,
+    _iot_device_module_create,
+    _iot_device_module_show,
+    _iot_device_create,
+    _iot_device_delete,
+    _iot_edge_set_modules,
+    _iot_hub_configuration_delete,
+    _iot_device_module_list,
+    _iot_device_twin_list,
+    _iot_hub_configuration_list,
+    _iot_hub_configuration_create,
+    _iot_device_twin_update,
+    _iot_device_module_twin_update,
+)
+from azext_iot._factory import iot_hub_service_factory
+from knack.prompting import prompt_y_n
+import json
+from tqdm import tqdm
+from typing import List, Optional
+import os
+
+logger = get_logger(__name__)
+cli = EmbeddedCLI()
+
+
+OVERWRITE_FILE_MSG = "File {0} is not empty. Overwrite file? "
+FILE_NOT_EMPTY_MSG = "Command aborted. Include the --replace flag to overwrite file."
+
+
+class StateProvider(IoTHubProvider):
+    def __init__(
+        self,
+        cmd,
+        hub: Optional[str] = None,
+        rg: Optional[str] = None,
+        login: Optional[str] = None,
+        auth_type_dataplane: Optional[str] = None,
+    ):
+        try:
+            super(StateProvider, self).__init__(
+                cmd=cmd,
+                hub_name=hub,
+                rg=rg,
+                login=login,
+                auth_type_dataplane=auth_type_dataplane
+            )
+        except ResourceNotFoundError:
+            self.target = None
+            self.resolver = None
+        self.auth_type = auth_type_dataplane
+
+        if not self.hub_name:
+            self.hub_name = self.target["entity"].split('.')[0]
+        if not self.rg and not self.target:
+            raise RequiredArgumentMissingError("Please provide the resource group for the hub that will be created.")
```

Comment:
This error does not make sense when attempting to export from a non-existent hub like so:

az iot hub state export --aspects configurations -f /file/location/state.json -n NonExistentHub